runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/grpcclient/GrpcGetWorkerMetadataStream.java

```diff
@@ -0,0 +1,177 @@
+    this.serverMappingUpdater = serverMappingUpdater;
+  }
+
+  public static GrpcGetWorkerMetadataStream create(
+      Function<StreamObserver<WorkerMetadataResponse>, StreamObserver<WorkerMetadataRequest>>
+          startGetWorkerMetadataRpcFn,
+      BackOff backoff,
+      StreamObserverFactory streamObserverFactory,
+      Set<AbstractWindmillStream<?, ?>> streamRegistry,
+      int logEveryNStreamFailures,
+      JobHeader jobHeader,
+      AtomicLong metadataVersion,
+      ThrottleTimer getWorkerMetadataThrottleTimer,
+      Consumer<WindmillEndpoints> serverMappingUpdater) {
+    GrpcGetWorkerMetadataStream getWorkerMetadataStream =
+        new GrpcGetWorkerMetadataStream(
+            startGetWorkerMetadataRpcFn,
+            backoff,
+            streamObserverFactory,
+            streamRegistry,
+            logEveryNStreamFailures,
+            jobHeader,
+            metadataVersion,
+            getWorkerMetadataThrottleTimer,
+            serverMappingUpdater);
+    getWorkerMetadataStream.startStream();
+    return getWorkerMetadataStream;
+  }
+
+  @VisibleForTesting
+  static GrpcGetWorkerMetadataStream forTesting(
+      Function<StreamObserver<WorkerMetadataResponse>, StreamObserver<WorkerMetadataRequest>>
+          getWorkerMetadataRpc,
+      int metadataVersion,
+      JobHeader jobHeader,
+      Set<AbstractWindmillStream<?, ?>> streamRegistry,
+      Consumer<WindmillEndpoints> serverMappingUpdater) {
+    GrpcGetWorkerMetadataStream testStream =
+        new GrpcGetWorkerMetadataStream(
+            getWorkerMetadataRpc,
+            FluentBackoff.DEFAULT.backoff(),
+            StreamObserverFactory.direct(DEFAULT_STREAM_RPC_DEADLINE_SECONDS * 2, 1),
+            streamRegistry,
+            1, // logEveryNStreamFailures
+            jobHeader,
+            new AtomicLong(metadataVersion),
+            new ThrottleTimer(),
+            serverMappingUpdater);
+    testStream.startStream();
+    return testStream;
+  }
+
+  private static long updateMetadataVersion(
+      long currentMetadataVersion, long responseMetadataVersion) {
+    if (currentMetadataVersion < responseMetadataVersion) {
+      return responseMetadataVersion;
+    }
+
+    // If the currentMetadataVersion is greater than or equal to one in the response, the response
+    // data is stale, and we do not want to do anything.
+    LOG.info(
+        "Received metadata version={}; Current metadata version={}. "
+            + "Skipping update because received stale metadata",
+        responseMetadataVersion,
+        currentMetadataVersion);
+    return currentMetadataVersion;
+  }
+
+  @Override
+  protected void onResponse(WorkerMetadataResponse response) {
+    long currentMetadataVersion = metadataVersion.get();
+    long updatedMetadataVersion =
+        metadataVersion.updateAndGet(
+            current -> updateMetadataVersion(current, response.getMetadataVersion()));
```

Comment:
how about inlining the non-logging part of updateMetadataVersion here, and moving the logging below to else? Seems clearer than logging in the separate method.

Reply:
done